File: exam-back/src/main/java/com/examback/member/dto/MemberDto.java
package com.examback.member.dto;

import lombok.*;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class MemberDto {
    private String memId;
    private LocalDateTime createDateTime;
    private LocalDateTime updateDateTime;
    private String username;
    private String password;
    private String nickname;
    private String memType;
    private Boolean useYn;
    private String role;
    public static MemberDto from(SignUpDto signUpDto) {
        return MemberDto.builder()
                .username(signUpDto.getUsername())
                .nickname(signUpDto.getNickname())
                .password(signUpDto.getPassword1())
                .memType(signUpDto.getMemType())
                .build();
    }


    public List<GrantedAuthority> getAuthorities() {
        List<GrantedAuthority> authorities = new ArrayList<>();
        if(this.getRole().equals("ADMIN")) {
            authorities.add(new SimpleGrantedAuthority("ADMIN"));
        } else if(this.getRole().equals("SUBADMIN")) {
            authorities.add(new SimpleGrantedAuthority("SUBADMIN"));
        } else {
            authorities.add(new SimpleGrantedAuthority("MEMBER"));
        }

        return authorities;
    }
}

File: kotlin-back/src/main/kotlin/com/example/kotlinback/member/dto/MemberDto.kt
package com.example.kotlinback.member.dto

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import java.time.LocalDate
import java.time.LocalDateTime

data class MemberDto(
    val memId: String = "",
    val createDateTime: LocalDateTime?,
    val updateDateTime: LocalDateTime?,
    val username: String = "",
    val email: String = "",
    val password: String = "",
    val nickname: String = "",
    val memType: String = "",
    val useYn: Boolean = true,
    val role: String = ""
) {
    val authorities: List<GrantedAuthority>
        get() {
            val authorities: MutableList<GrantedAuthority> = ArrayList()
            if (this.role == "ADMIN") {
                authorities.add(SimpleGrantedAuthority("ADMIN"))
            } else if (this.role == "SUBADMIN") {
                authorities.add(SimpleGrantedAuthority("SUBADMIN"))
            } else {
                authorities.add(SimpleGrantedAuthority("MEMBER"))
            }
            return authorities
        }

    companion object {
        fun from(signUpDto: SignUpDto): MemberDto {
            return MemberDto(
                username = signUpDto.username,
                nickname = signUpDto.nickname,
                password = signUpDto.password1,
                memType = signUpDto.memType,
                email = signUpDto.email,
                createDateTime = LocalDateTime.now(),
                updateDateTime = LocalDateTime.now()
            )
        }
    }
}
